### PR TITLE
Pr/add-tls13-moving-state-tests

### PR DESCRIPTION
**Pull request by yuhaoth**
## Description
Add state moving test for TLS1.3

preceding PR: #5749 

## Status
**IN DEVELOPMENT**

### CODE REVIEW

**Comment by ronald-cron-arm** [comment]
@yuhaoth now that #5749 is merged, please rebase to clarify what belongs to this PR and what does not.

**Comment by yuhaoth** [comment]
rebased

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: tests/suites/test_suite_ssl.function [hunk 1]
```diff
@@ -2096,6 +2099,17 @@ void perform_handshake( handshake_test_options* options )
     }
 
     TEST_ASSERT( mbedtls_ssl_is_handshake_over( &client.ssl ) == 1 );
+
+    /* Workaround for TLS1.3 */
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply tests/suites/test_suite_ssl.function:2103]: Will update it.
When client.ssl reach handshake_over state, server.ssl does not reach handshake_over state then next assert will raise error

**Comment by yuhaoth** [comment]
Rebased . And not address feedback

File: tests/suites/test_suite_ssl.function (hunk 1 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply tests/suites/test_suite_ssl.function:2103]: Updated. I think we should make sure server.ssl also before check it. 
The origin code assume server.ssl is HANDSHAKE_OVER when client.ssl is HANDSHAKE_OVER.
I think it is wrong.

File: tests/suites/test_suite_ssl.data [hunk 2]
```diff
@@ -105,153 +105,285 @@ depends_on:MBEDTLS_SSL_PROTO_TLS1_2
 mbedtls_endpoint_sanity:MBEDTLS_SSL_IS_SERVER
 
 Test moving clients handshake to state: HELLO_REQUEST
-depends_on:MBEDTLS_SSL_PROTO_TLS1_2
+depends_on:MBEDTLS_SSL_PROTO_TLS1_2:!MBEDTLS_SSL_PROTO_TLS1_3
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply tests/suites/test_suite_ssl.data]: @ronald-cron-arm , I just removed all TLS1.3 dependency , and I found CERTIFICATE_REQUEST of servers fail.  Could you check if it is reasonable ? The fail will be shown in #5357

I'd like remove the dependency step by step. I will add more change tomorrow.

[the remark this answers is not in the document]
Comment (yuhaoth) [reply tests/suites/test_suite_ssl.data]: Done. please review

File: tests/suites/test_suite_ssl.function [hunk 3]
```diff
@@ -913,6 +914,8 @@ int mbedtls_endpoint_init( mbedtls_endpoint *ep, int endpoint_type, int pk_alg,
         &( ep->ctr_drbg ) );
     mbedtls_entropy_init( &( ep->entropy ) );
 
+
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_ssl.function:917]:
```suggestion
```

File: tests/suites/test_suite_ssl.function [hunk 4]
```diff
@@ -913,6 +914,8 @@ int mbedtls_endpoint_init( mbedtls_endpoint *ep, int endpoint_type, int pk_alg,
...
+
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_ssl.function:918]:
```suggestion
```

File: tests/suites/test_suite_ssl.data [hunk 5]
```diff
@@ -248,10 +228,31 @@ Negative test moving servers ssl to state: NEW_SESSION_TICKET
 depends_on:MBEDTLS_SSL_PROTO_TLS1_2
 move_handshake_to_state:MBEDTLS_SSL_IS_SERVER:MBEDTLS_SSL_SERVER_NEW_SESSION_TICKET:0
 
+
+TLSv1.3:Test moving clients handshake to state: ENCRYPTED_EXTENSIONS
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_ssl.data:232]:
```suggestion
TLS 1.3:Test moving clients handshake to state: ENCRYPTED_EXTENSIONS
```
like in ssl-opt.sh. Please change all instances.

File: tests/suites/test_suite_ssl.data [hunk 6]
```diff
@@ -141,59 +136,49 @@ depends_on:MBEDTLS_SSL_PROTO_TLS1_2
 move_handshake_to_state:MBEDTLS_SSL_IS_CLIENT:MBEDTLS_SSL_CLIENT_KEY_EXCHANGE:1
 
 Test moving clients handshake to state: CERTIFICATE_VERIFY
-depends_on:MBEDTLS_SSL_PROTO_TLS1_2
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_ssl.data:144]: Just above (line 136), CLIENT_CERTIFICATE is a TLS 1.3 state on client side as well thus the test should not depend on MBEDTLS_SSL_PROTO_TLS1_2 I'd say. Ditto for FLUSH_BUFFERS (client and server), CLIENT_CERTIFICATE on server side below.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: tests/suites/test_suite_ssl.data (hunk 6 above)
Comment (yuhaoth) [reply tests/suites/test_suite_ssl.data:144]: `CLIENT_CERTIFICATE` is also used by TLS1.2 . I am not sure what's the problem.

**Comment by ronald-cron-arm** [comment]
The changes to get to the point where the "move to handshake state" tests are the same for TLS 1.2 and TLS 1.3 when the state is shared between TLS 1.2 and TLS 1.3 were a bit tricky to explain in review comments thus I pushed three additional commits instead. @yuhaoth, please have a look.

**Review by paul-elliott-arm — COMMENTED** [review]
Mainly commenting so that Github remembers state, but I have reviewed up to this point, and it all seems good so far. I find Ronald's change to the logic makes things a lot easier to understand.

**Comment by yuhaoth** [comment]
> The changes to get to the point where the "move to handshake state" tests are the same for TLS 1.2 and TLS 1.3 when the state is shared between TLS 1.2 and TLS 1.3 were a bit tricky to explain in review comments thus I pushed three additional commits instead. @yuhaoth, please have a look.

The changes look good.

**Comment by ronald-cron-arm** [comment]
@yuhaoth @paul-elliott-arm I've just pushed a last commit that adds a change log for the bug fixed in this PR. I approve the PR up to my commits. Then we need the approval of @yuhaoth for my commit and @paul-elliott-arm for the whole PR and we can then merge.

**Review by yuhaoth — COMMENTED** [review]
LGTM. But I can not approve my PR :)

**Pull request merged**